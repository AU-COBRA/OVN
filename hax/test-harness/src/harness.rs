mod command_hax_ext;
use command_hax_ext::*;
use serde_json::{Map, Value};
use std::process::{Command, Stdio};

#[derive(Clone, Debug, serde::Serialize)]
pub enum TestKind {
    Translate { backend: String },
}

impl TestKind {
    fn as_name(&self) -> String {
        (match self {
            TestKind::Translate { backend } => ["into".to_string(), backend.clone()],
        })
        .join("-")
    }
}

#[allow(dead_code)]
fn bool_true() -> bool {
    true
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct TestSnapshot {
    #[serde(default = "bool_true")]
    pub stderr: bool,
    #[serde(default = "bool_true")]
    pub stdout: bool,
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct TestSpec {
    /// is the test optional? (useful for slow tests for instance)
    pub optional: bool,
    /// a broken test a test that should succeed (or fail) but does
    /// not dues to a bug to be fixed (see field [issue_id] below)
    pub broken: bool,
    /// Github issue ID
    pub issue_id: Option<u64>,
    /// Is that a positive or a negative test?
    pub positive: bool,
    pub snapshot: TestSnapshot,
    pub include_flag: Option<String>,
    pub backend_options: Option<Vec<String>>,
}

impl From<Value> for TestSpec {
    /// Parse a JSON value into a TestSpec
    fn from(o: Value) -> Self {
        fn as_opt_bool(v: &Value, def: bool) -> Option<bool> {
            if v.is_null() {
                return Some(def);
            }
            v.as_bool()
        }
        fn as_bool(o: &Value, k: &str, def: bool) -> bool {
            let v = &o[k];
            as_opt_bool(v, def)
                .expect(format!("[{}] was expected to be a boolean, got {}", k, v).as_str())
        }
        let snapshot = &o["snapshot"];
        TestSpec {
            optional: as_bool(&o, "optional", false),
            broken: as_bool(&o, "broken", false),
            positive: as_bool(&o, "positive", true),
            issue_id: o["positive"].as_u64(),
            include_flag: o["include-flag"].as_str().map(|s| s.into()),
            backend_options: serde_json::from_value(o["backend-options"].clone()).unwrap(),
            snapshot: as_opt_bool(snapshot, true)
                .map(|b| TestSnapshot {
                    stderr: b,
                    stdout: b,
                })
                .or_else(|| match snapshot.as_str() {
                    Some(v @ ("stdout" | "stderr" | "both" | "none")) => Some(TestSnapshot {
                        stdout: matches!(v, "stdout" | "both"),
                        stderr: matches!(v, "stderr" | "both"),
                    }),
                    Some(v) => panic!(
                        "[snapshot] is \"{}\" but was expected to be \"stderr\", \"stdout\" or \"both\"", v
                    ),
                    None => None,
                })
                .unwrap_or_else(|| TestSnapshot {
                    stderr: as_bool(&snapshot, "stderr", true),
                    stdout: as_bool(&snapshot, "stdout", true),
                }),
        }
    }
}

/// The information for a test is given by `cargo metadata`
#[derive(Clone, Debug, serde::Serialize)]
pub struct TestInfo {
    pub name: String,
    pub manifest: std::path::PathBuf,
    pub description: Option<String>,
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct Test {
    pub kind: TestKind,
    pub info: TestInfo,
    pub spec: TestSpec,
}

impl Test {
    fn as_args(&self) -> Vec<String> {
        match &self.kind {
            TestKind::Translate { backend } => {
                let mut args = vec![];
                args.push("into".to_string());
                if let Some(i) = self.spec.include_flag.as_ref() {
                    args.push("-i".to_string());
                    args.push(i.to_string());
                }
                args.push("--dry-run".to_string());
                args.push(backend.clone());
                if let Some(backend_options) = &self.spec.backend_options {
                    args.extend_from_slice(backend_options.clone().as_slice());
                }
                args
            }
        }
    }
}

impl std::fmt::Display for Test {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{} - {:?}", self.info.name, self.kind)?;
        if let Some(issue_id) = self.spec.issue_id {
            write!(f, " #{}", issue_id)?;
        };
        Ok(())
    }
}

impl Test {
    fn into_runner(self, workspace: String) -> Result<(), libtest_mimic::Failed> {
        // 1. cook a command
        let mut cmd = Command::hax(&["-C"]);
        cmd.arg("--manifest-path").arg(self.info.manifest.clone());
        cmd.arg(";");
        cmd.stdout(Stdio::piped()).stderr(Stdio::piped());
        cmd.args(self.as_args());

        // 2. execute it (twice, idea of @franziskuskiefer, so that
        // the messages related to building dependencies are not
        // included in the second one)
        let _ = cmd.output().unwrap();
        let out = cmd.output().unwrap();

        let command_successful = out.status.success();
        let cleanup = |s: String| {
            use lazy_static::lazy_static;
            use regex::Regex;
            lazy_static! {
                // Regex [TIME] matches compile times
                static ref TIME: Regex = Regex::new(r"\bin \d+(\.\d+)?s\b").unwrap();
                static ref LOCK: Regex = Regex::new(r"Blocking waiting for \w+ lock on (the registry index|build directory|package cache)").unwrap();
            }
            TIME.replace_all(
                LOCK.replace_all(
                    &s.replace(r"\", "/").replace(&workspace, "WORKSPACE_ROOT"),
                    "",
                )
                .as_ref(),
                "in XXs",
            )
            .trim()
            .to_string()
        };
        let serr = cleanup(String::from_utf8_lossy(&out.stderr).to_string());
        let sout = String::from_utf8_lossy(&out.stdout).to_string();

        // 3. make sure the test is successful
        let mut snapshot: Map<String, Value> = Map::new();
        if self.spec.snapshot.stderr {
            snapshot.insert("stderr".to_string(), Value::String(serr.clone()));
        }
        if self.spec.snapshot.stdout {
            snapshot.insert(
                "stdout".to_string(),
                serde_json::from_str(&sout)
                    .map(|out: hax_types::engine_api::Output| {
                        use serde_json::json;
                        json!({
                            "diagnostics": Value::Array(out.diagnostics.into_iter().map(|diag| json!({
                                "spans": Value::Array(diag.span.clone().into_iter().map(|span| Value::String(format!("{:?}", span))).collect()),
                                "message": Value::String(format!("{}", diag)),
                            })).collect()),
                            "files": Value::Object(out.files.into_iter().map(|file| (file.path, Value::String(file.contents))).collect())
                        })
                    })
                    .unwrap_or_else(|_| Value::String(cleanup(sout.clone()))),
            );
        }

        if !snapshot.is_empty() {
            let exit = out.status.code().unwrap_or(std::i32::MAX);
            snapshot.insert("exit".to_string(), exit.into());
            let snapshot = Value::Object(snapshot);
            let name = format!("{} {}", self.info.name, self.kind.as_name());

            let mut info = self.clone();
            info.info.manifest = info.info.manifest.strip_prefix(workspace).unwrap().into();
            insta::with_settings!({
                info => &info,
            }, { insta::assert_toml_snapshot!(name, snapshot) })
        }

        let err = |s: &str| {
            Err(format!(
                "Command {s}.\nThe command was: {:?}{}",
                cmd,
                if command_successful {
                    "".to_string()
                } else {
                    format!("\nSTDOUT:\n{}\nSTDERR:\n{}", sout, serr)
                }
            ))
        };
        match (command_successful, (self.spec.positive, self.spec.broken)) {
            (true, (true, false) | (false, true)) => Ok(()),
            (false, (false, false) | (true, true)) => Ok(()),
            (false, (false, true)) => err("failed, but this is a negative test marked broken")?,
            (false, (true, false)) => err("failed")?,
            (true, (true, true)) => err("succeeded, but this is a positive test marked broken")?,
            (true, (false, false)) => err("succeeded, but this is a negative test")?,
        }
    }

    fn into_trial(&self, workspace: &String) -> libtest_mimic::Trial {
        libtest_mimic::Trial::test(format!("{}", &self), {
            let test = self.clone();
            let workspace = workspace.clone();
            move || test.clone().into_runner(workspace)
        })
        .with_kind(if self.spec.positive {
            "positive"
        } else {
            "negative"
        })
        .with_ignored_flag(self.spec.optional)
    }
}

/// Given [metadata] the table declared in a test's [Cargo.toml]
/// [workspace.hax-tests], this function returns a list of tests
fn parse_hax_tests_metadata(info: TestInfo, metadata: &Value) -> Vec<Test> {
    if metadata.is_null() {
        return vec![];
    }

    metadata
        .as_object()
        .expect(
            format!(
                "Expected value at key [hax-tests] to be a dictionary for package {:#?}",
                info
            )
            .as_str(),
        )
        .into_iter()
        .flat_map(|(a, o)| {
            o.as_object()
                .expect(
                    format!(
                        "Expected value at key [{}] be a dictionary for package {:#?}",
                        a, info
                    )
                    .as_str(),
                )
                .into_iter()
                .flat_map(|(key, o)| key.split("+").map(|k| (k.trim().to_string(), o.clone())))
                .map(|(b, o)| (a.clone(), b, o))
        })
        .map(|(a, b, o)| Test {
            spec: o.into(),
            info: info.clone(),
            kind: match a.as_str() {
                "into" => TestKind::Translate { backend: b },
                _ => panic!(
                    "unexpected metadata [hax-tests.{}.{}] for package {:#?}",
                    a, b, info
                ),
            },
        })
        .collect()
}

fn main() {
    let metadata = cargo_metadata::MetadataCommand::new()
        .manifest_path("../tests/Cargo.toml")
        .exec()
        .unwrap();
    let workspace_root: String = metadata.workspace_root.into();

    let mut args = libtest_mimic::Arguments::from_args();
    args.test_threads = Some(1);

    libtest_mimic::run(
        &args,
        metadata
            .packages
            .into_iter()
            .flat_map(|o| {
                parse_hax_tests_metadata(
                    TestInfo {
                        name: o.name,
                        description: o.description,
                        manifest: o.manifest_path.into(),
                    },
                    &o.metadata["hax-tests"],
                )
            })
            .map(|test| test.into_trial(&workspace_root))
            .collect(),
    )
    .exit();
}
