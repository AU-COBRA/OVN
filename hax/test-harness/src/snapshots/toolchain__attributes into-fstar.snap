---
source: test-harness/src/harness.rs
expression: snapshot
info:
  kind:
    Translate:
      backend: fstar
  info:
    name: attributes
    manifest: attributes/Cargo.toml
    description: ~
  spec:
    optional: false
    broken: false
    issue_id: ~
    positive: true
    snapshot:
      stderr: false
      stdout: true
    include_flag: ~
    backend_options: ~
---
exit = 0

[stdout]
diagnostics = []

[stdout.files]
"Attributes.Ensures_on_arity_zero_fns.fst" = '''
module Attributes.Ensures_on_arity_zero_fns
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

let basically_a_constant (_: Prims.unit)
    : Prims.Pure u8
      (requires true)
      (ensures
        fun x ->
          let x:u8 = x in
          x >. 100uy) = 127uy

let doing_nothing (_: Prims.unit)
    : Prims.Pure Prims.unit
      (requires true)
      (ensures
        fun v__x ->
          let v__x:Prims.unit = v__x in
          true) = ()
'''
"Attributes.Inlined_code_ensures_requires.fst" = '''
module Attributes.Inlined_code_ensures_requires
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

let increment_array (v: t_Array u8 (sz 4))
    : Prims.Pure (t_Array u8 (sz 4))
      (requires forall i. FStar.Seq.index v i <. 254uy)
      (ensures
        fun vv_future ->
          let vv_future:t_Array u8 (sz 4) = vv_future in
          let future_v:t_Array u8 (sz 4) = vv_future in
          forall i. FStar.Seq.index future_v i >. 0uy) =
  let v:t_Array u8 (sz 4) =
    Rust_primitives.Hax.Monomorphized_update_at.update_at_usize v
      (sz 0)
      ((v.[ sz 0 ] <: u8) +! 1uy <: u8)
  in
  let v:t_Array u8 (sz 4) =
    Rust_primitives.Hax.Monomorphized_update_at.update_at_usize v
      (sz 1)
      ((v.[ sz 1 ] <: u8) +! 1uy <: u8)
  in
  let v:t_Array u8 (sz 4) =
    Rust_primitives.Hax.Monomorphized_update_at.update_at_usize v
      (sz 2)
      ((v.[ sz 2 ] <: u8) +! 1uy <: u8)
  in
  let v:t_Array u8 (sz 4) =
    Rust_primitives.Hax.Monomorphized_update_at.update_at_usize v
      (sz 3)
      ((v.[ sz 3 ] <: u8) +! 1uy <: u8)
  in
  v
'''
"Attributes.Int_model.fst" = '''
module Attributes.Int_model
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

unfold type t_Int = int

unfold let add x y = x + y

unfold instance impl: Core.Ops.Arith.t_Sub t_Int t_Int =
  {
    f_Output = t_Int;
    f_sub_pre = (fun (self: t_Int) (other: t_Int) -> true);
    f_sub_post = (fun (self: t_Int) (other: t_Int) (out: t_Int) -> true);
    f_sub = fun (self: t_Int) (other: t_Int) -> self + other
  }
'''
"Attributes.Nested_refinement_elim.fst" = '''
module Attributes.Nested_refinement_elim
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

let t_DummyRefinement = x: u16{true}

let elim_twice (x: t_DummyRefinement) : u16 = ((x <: u16) <: t_DummyRefinement) <: u16
'''
"Attributes.Newtype_pattern.fst" = '''
module Attributes.Newtype_pattern
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

let v_MAX: usize = sz 10

type t_SafeIndex = { f_i:f_i: usize{f_i <. v_MAX} }

let impl__SafeIndex__as_usize (self: t_SafeIndex) : usize = self.f_i

let impl__SafeIndex__new (i: usize) : Core.Option.t_Option t_SafeIndex =
  if i <. v_MAX
  then Core.Option.Option_Some ({ f_i = i } <: t_SafeIndex) <: Core.Option.t_Option t_SafeIndex
  else Core.Option.Option_None <: Core.Option.t_Option t_SafeIndex

[@@ FStar.Tactics.Typeclasses.tcinstance]
let impl_1 (#v_T: Type0) : Core.Ops.Index.t_Index (t_Array v_T (sz 10)) t_SafeIndex =
  {
    f_Output = v_T;
    f_index_pre = (fun (self: t_Array v_T (sz 10)) (index: t_SafeIndex) -> true);
    f_index_post = (fun (self: t_Array v_T (sz 10)) (index: t_SafeIndex) (out: v_T) -> true);
    f_index = fun (self: t_Array v_T (sz 10)) (index: t_SafeIndex) -> self.[ index.f_i ]
  }
'''
"Attributes.Pre_post_on_traits_and_impls.fst" = '''
module Attributes.Pre_post_on_traits_and_impls
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

class t_Operation (v_Self: Type0) = {
  f_double_pre:x: u8
    -> pred:
      Type0
        { (Rust_primitives.Hax.Int.from_machine x <: Hax_lib.Int.t_Int) <=
          (127 <: Hax_lib.Int.t_Int) ==>
          pred };
  f_double_post:x: u8 -> result: u8
    -> pred:
      Type0
        { pred ==>
          ((Rust_primitives.Hax.Int.from_machine x <: Hax_lib.Int.t_Int) * (2 <: Hax_lib.Int.t_Int)
            <:
            Hax_lib.Int.t_Int) =
          (Rust_primitives.Hax.Int.from_machine result <: Hax_lib.Int.t_Int) };
  f_double:x0: u8 -> Prims.Pure u8 (f_double_pre x0) (fun result -> f_double_post x0 result)
}

class t_TraitWithRequiresAndEnsures (v_Self: Type0) = {
  f_method_pre:self___: v_Self -> x: u8 -> pred: Type0{x <. 100uy ==> pred};
  f_method_post:self___: v_Self -> x: u8 -> r: u8 -> pred: Type0{pred ==> r >. 88uy};
  f_method:x0: v_Self -> x1: u8
    -> Prims.Pure u8 (f_method_pre x0 x1) (fun result -> f_method_post x0 x1 result)
}

let test
      (#v_T: Type0)
      (#[FStar.Tactics.Typeclasses.tcresolve ()] i1: t_TraitWithRequiresAndEnsures v_T)
      (x: v_T)
    : u8 = (f_method #v_T #FStar.Tactics.Typeclasses.solve x 99uy <: u8) -! 88uy

type t_ViaAdd = | ViaAdd : t_ViaAdd

[@@ FStar.Tactics.Typeclasses.tcinstance]
let impl: t_Operation t_ViaAdd =
  {
    f_double_pre
    =
    (fun (x: u8) ->
        (Rust_primitives.Hax.Int.from_machine x <: Hax_lib.Int.t_Int) <= (127 <: Hax_lib.Int.t_Int));
    f_double_post
    =
    (fun (x: u8) (result: u8) ->
        ((Rust_primitives.Hax.Int.from_machine x <: Hax_lib.Int.t_Int) * (2 <: Hax_lib.Int.t_Int)
          <:
          Hax_lib.Int.t_Int) =
        (Rust_primitives.Hax.Int.from_machine result <: Hax_lib.Int.t_Int));
    f_double = fun (x: u8) -> x +! x
  }

type t_ViaMul = | ViaMul : t_ViaMul

[@@ FStar.Tactics.Typeclasses.tcinstance]
let impl_1: t_Operation t_ViaMul =
  {
    f_double_pre
    =
    (fun (x: u8) ->
        (Rust_primitives.Hax.Int.from_machine x <: Hax_lib.Int.t_Int) <= (127 <: Hax_lib.Int.t_Int));
    f_double_post
    =
    (fun (x: u8) (result: u8) ->
        ((Rust_primitives.Hax.Int.from_machine x <: Hax_lib.Int.t_Int) * (2 <: Hax_lib.Int.t_Int)
          <:
          Hax_lib.Int.t_Int) =
        (Rust_primitives.Hax.Int.from_machine result <: Hax_lib.Int.t_Int));
    f_double = fun (x: u8) -> x *! 2uy
  }
'''
"Attributes.Refined_arithmetic.fst" = '''
module Attributes.Refined_arithmetic
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

type t_Foo = | Foo : u8 -> t_Foo

[@@ FStar.Tactics.Typeclasses.tcinstance]
let impl: Core.Ops.Arith.t_Add t_Foo t_Foo =
  {
    f_Output = t_Foo;
    f_add_pre = (fun (self___: t_Foo) (rhs: t_Foo) -> self___._0 <. (255uy -! rhs._0 <: u8));
    f_add_post = (fun (self: t_Foo) (rhs: t_Foo) (out: t_Foo) -> true);
    f_add = fun (self: t_Foo) (rhs: t_Foo) -> Foo (self._0 +! rhs._0) <: t_Foo
  }

[@@ FStar.Tactics.Typeclasses.tcinstance]
let impl_1: Core.Ops.Arith.t_Mul t_Foo t_Foo =
  {
    f_Output = t_Foo;
    f_mul_pre
    =
    (fun (self___: t_Foo) (rhs: t_Foo) -> rhs._0 =. 0uy || self___._0 <. (255uy /! rhs._0 <: u8));
    f_mul_post = (fun (self: t_Foo) (rhs: t_Foo) (out: t_Foo) -> true);
    f_mul = fun (self: t_Foo) (rhs: t_Foo) -> Foo (self._0 *! rhs._0) <: t_Foo
  }
'''
"Attributes.Refined_indexes.fst" = '''
module Attributes.Refined_indexes
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

let v_MAX: usize = sz 10

type t_MyArray = | MyArray : t_Array u8 (sz 10) -> t_MyArray

/// Triple dash comment
(** Multiline double star comment Maecenas blandit accumsan feugiat.
    Done vitae ullamcorper est.
    Curabitur id dui eget sem viverra interdum. *)
let mutation_example
      (use_generic_update_at: t_MyArray)
      (use_specialized_update_at: t_Slice u8)
      (specialized_as_well: Alloc.Vec.t_Vec u8 Alloc.Alloc.t_Global)
    : (t_MyArray & t_Slice u8 & Alloc.Vec.t_Vec u8 Alloc.Alloc.t_Global) =
  let use_generic_update_at:t_MyArray =
    Rust_primitives.Hax.update_at use_generic_update_at (sz 2) 0uy
  in
  let use_specialized_update_at:t_Slice u8 =
    Rust_primitives.Hax.Monomorphized_update_at.update_at_usize use_specialized_update_at (sz 2) 0uy
  in
  let specialized_as_well:Alloc.Vec.t_Vec u8 Alloc.Alloc.t_Global =
    Rust_primitives.Hax.Monomorphized_update_at.update_at_usize specialized_as_well (sz 2) 0uy
  in
  use_generic_update_at, use_specialized_update_at, specialized_as_well
  <:
  (t_MyArray & t_Slice u8 & Alloc.Vec.t_Vec u8 Alloc.Alloc.t_Global)

[@@ FStar.Tactics.Typeclasses.tcinstance]
let impl: Core.Ops.Index.t_Index t_MyArray usize =
  {
    f_Output = u8;
    f_index_pre = (fun (self___: t_MyArray) (index: usize) -> index <. v_MAX);
    f_index_post = (fun (self: t_MyArray) (index: usize) (out: u8) -> true);
    f_index = fun (self: t_MyArray) (index: usize) -> self.[ index ]
  }
'''
"Attributes.Refinement_types.fst" = '''
module Attributes.Refinement_types
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

let t_BoundedU8 (v_MIN v_MAX: u8) = x: u8{x >=. v_MIN && x <=. v_MAX}

/// Example of a specific constraint on a value
let t_CompressionFactor = x: u8{x =. 4uy || x =. 5uy || x =. 10uy || x =. 11uy}

/// Even `u8` numbers. Constructing pub Even values triggers static
/// proofs in the extraction.
let t_Even = x: u8{(x %! 2uy <: u8) =. 0uy}

/// A field element
let t_FieldElement = x: u16{x <=. 2347us}

/// A modular mutliplicative inverse
let t_ModInverse (v_MOD: u32) =
  n:
  u32
    { (((cast (n <: u32) <: u128) *! (cast (v_MOD <: u32) <: u128) <: u128) %!
        (cast (v_MOD <: u32) <: u128)
        <:
        u128) =.
      pub_u128 1 }

/// A string that contains no space.
let t_NoE =
  x:
  Alloc.String.t_String
    { let _, out:(Core.Str.Iter.t_Chars & bool) =
        Core.Iter.Traits.Iterator.f_any #Core.Str.Iter.t_Chars
          #FStar.Tactics.Typeclasses.solve
          (Core.Str.impl__str__chars (Core.Ops.Deref.f_deref #Alloc.String.t_String
                  #FStar.Tactics.Typeclasses.solve
                  x
                <:
                string)
            <:
            Core.Str.Iter.t_Chars)
          (fun ch ->
              let ch:char = ch in
              ch =. ' ' <: bool)
      in
      ~.out }

let bounded_u8 (x: t_BoundedU8 12uy 15uy) (y: t_BoundedU8 10uy 11uy) : t_BoundedU8 1uy 23uy =
  (x <: u8) +! (y <: u8) <: t_BoundedU8 1uy 23uy

let double (x: u8) : Prims.Pure t_Even (requires x <. 127uy) (fun _ -> Prims.l_True) =
  x +! x <: t_Even

let double_refine (x: u8) : Prims.Pure t_Even (requires x <. 127uy) (fun _ -> Prims.l_True) =
  x +! x <: t_Even
'''
"Attributes.Requires_mut.fst" = '''
module Attributes.Requires_mut
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

class t_Foo (v_Self: Type0) = {
  f_f_pre:x: u8 -> y: u8
    -> pred:
      Type0
        { ((Rust_primitives.Hax.Int.from_machine x <: Hax_lib.Int.t_Int) +
            (Rust_primitives.Hax.Int.from_machine y <: Hax_lib.Int.t_Int)
            <:
            Hax_lib.Int.t_Int) <
          (254 <: Hax_lib.Int.t_Int) ==>
          pred };
  f_f_post:x: u8 -> y: u8 -> x1: (u8 & u8)
    -> pred:
      Type0
        { pred ==>
          (let y_future, output_variable:(u8 & u8) = x1 in
            output_variable =. y_future) };
  f_f:x0: u8 -> x1: u8 -> Prims.Pure (u8 & u8) (f_f_pre x0 x1) (fun result -> f_f_post x0 x1 result);
  f_g_pre:u8 -> u8 -> Type0;
  f_g_post:u8 -> u8 -> u8 -> Type0;
  f_g:x0: u8 -> x1: u8 -> Prims.Pure u8 (f_g_pre x0 x1) (fun result -> f_g_post x0 x1 result);
  f_h_pre:u8 -> u8 -> Type0;
  f_h_post:u8 -> u8 -> Prims.unit -> Type0;
  f_h:x0: u8 -> x1: u8
    -> Prims.Pure Prims.unit (f_h_pre x0 x1) (fun result -> f_h_post x0 x1 result);
  f_i_pre:u8 -> u8 -> Type0;
  f_i_post:u8 -> u8 -> u8 -> Type0;
  f_i:x0: u8 -> x1: u8 -> Prims.Pure u8 (f_i_pre x0 x1) (fun result -> f_i_post x0 x1 result)
}

[@@ FStar.Tactics.Typeclasses.tcinstance]
let impl: t_Foo Prims.unit =
  {
    f_f_pre
    =
    (fun (x: u8) (y: u8) ->
        ((Rust_primitives.Hax.Int.from_machine x <: Hax_lib.Int.t_Int) +
          (Rust_primitives.Hax.Int.from_machine y <: Hax_lib.Int.t_Int)
          <:
          Hax_lib.Int.t_Int) <
        (254 <: Hax_lib.Int.t_Int));
    f_f_post
    =
    (fun (x: u8) (y: u8) (y_future, output_variable: (u8 & u8)) -> output_variable =. y_future);
    f_f
    =
    (fun (x: u8) (y: u8) ->
        let y:u8 = y +! x in
        let hax_temp_output:u8 = y in
        y, hax_temp_output <: (u8 & u8));
    f_g_pre = (fun (x: u8) (y: u8) -> true);
    f_g_post = (fun (x: u8) (y: u8) (output_variable: u8) -> output_variable =. y);
    f_g = (fun (x: u8) (y: u8) -> y);
    f_h_pre = (fun (x: u8) (y: u8) -> true);
    f_h_post
    =
    (fun (x: u8) (y: u8) (output_variable: Prims.unit) -> output_variable =. (() <: Prims.unit));
    f_h = (fun (x: u8) (y: u8) -> () <: Prims.unit);
    f_i_pre = (fun (x: u8) (y: u8) -> true);
    f_i_post = (fun (x: u8) (y: u8) (y_future: u8) -> y_future =. y);
    f_i
    =
    fun (x: u8) (y: u8) ->
      let hax_temp_output:Prims.unit = () <: Prims.unit in
      y
  }
'''
"Attributes.Verifcation_status.fst" = '''
module Attributes.Verifcation_status
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

let a_panicfree_function (_: Prims.unit)
    : Prims.Pure u8
      Prims.l_True
      (ensures
        fun x ->
          let x:u8 = x in
          false) =
  let a:u8 = 3uy in
  let b:u8 = 6uy in
  let result:u8 = a +! b in
  let _:Prims.unit = admit () (* Panic freedom *) in
  result

let another_panicfree_function (_: Prims.unit)
    : Prims.Pure Prims.unit
      Prims.l_True
      (ensures
        fun x ->
          let x:Prims.unit = x in
          false) =
  let not_much:i32 = 0l in
  let nothing:i32 = 0l in
  let still_not_much:i32 = not_much +! nothing in
  admit () (* Panic freedom *)

#push-options "--admit_smt_queries true"

let a_function_which_only_laxes (_: Prims.unit) : Prims.unit = Hax_lib.v_assert false

#pop-options
'''
"Attributes.fst" = '''
module Attributes
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

let add3_lemma (x: u32)
    : Lemma Prims.l_True
      (ensures
        x <=. 10ul || x >=. (u32_max /! 3ul <: u32) || (add3 x x x <: u32) =. (x *! 3ul <: u32)) =
  ()

let inlined_code__V: u8 = 12uy

let issue_844_ (v__x: u8)
    : Prims.Pure u8
      Prims.l_True
      (ensures
        fun v__x_future ->
          let v__x_future:u8 = v__x_future in
          true) = v__x

let u32_max: u32 = 90000ul

unfold let some_function _ = "hello from F*"

/// A doc comment on `add3`
///another doc comment on add3
let add3 (x y z: u32)
    : Prims.Pure u32
      (requires x >. 10ul && y >. 10ul && z >. 10ul && ((x +! y <: u32) +! z <: u32) <. u32_max)
      (ensures
        fun result ->
          let result:u32 = result in
          Hax_lib.implies true
            (fun temp_0_ ->
                let _:Prims.unit = temp_0_ in
                result >. 32ul <: bool)) = (x +! y <: u32) +! z

let swap_and_mut_req_ens (x y: u32)
    : Prims.Pure (u32 & u32 & u32)
      (requires x <. 40ul && y <. 300ul)
      (ensures
        fun temp_0_ ->
          let x_future, y_future, result:(u32 & u32 & u32) = temp_0_ in
          x_future =. y && y_future =. x && result =. (x +! y <: u32)) =
  let x0:u32 = x in
  let x:u32 = y in
  let y:u32 = x0 in
  let hax_temp_output:u32 = x +! y in
  x, y, hax_temp_output <: (u32 & u32 & u32)

type t_Foo = {
  f_x:u32;
  f_y:f_y: u32{f_y >. 3ul};
  f_z:f_z: u32{((f_y +! f_x <: u32) +! f_z <: u32) >. 3ul}
}

let before_inlined_code = "example before"

let inlined_code (foo: t_Foo) : Prims.unit =
  let vv_a:i32 = 13l in
  let _:Prims.unit =
    let x = foo.f_x in
    let { f_y = y } = foo in
    add3 ((fun _ -> 3ul) foo) vv_a inlined_code__V y
  in
  ()

let inlined_code_after = "example after"
'''
